src/compiler/checker/lines_12000_16000.rs

```diff
@@ -390,7 +400,21 @@ impl TypeChecker {
             if false {
                 unimplemented!()
             } else {
-                type_set.push(type_.type_wrapper());
+                let len = type_set.len();
+                let index: isize = if len > 0 && type_.id() > type_set[len - 1].id() {
+                    !isize::try_from(len).unwrap()
+                } else {
+                    binary_search_copy_key(
+                        type_set,
+                        &type_.type_wrapper(),
+                        |type_, _| Some(self.get_type_id(type_)),
+                        compare_values,
+                        None,
+                    )
+                };
+                if index < 0 {
+                    type_set.insert(usize::try_from(!index).unwrap(), type_.type_wrapper());
```

Comment:
There's a lot of weird conversions between `isize` and `usize` because it wants to return negative values to mean something that should then be bit-negated (`~` in most languages, `!` in Rust) but then that value gets used as an index (which has to be a `usize` afaik)

Reply:
What does the bit-negation do? (Why would you negate in index?)